enhancements/network/static-ip-addresses-vsphere.md

```diff
@@ -0,0 +1,568 @@
+
+~~~mermaid
+sequenceDiagram
+    machineset controller->>+machine: creates machine with<br> preCreate hook
+    machine controller-->machine controller: waits for preCreate hook<br>to be removed
+    IP controller-->>+machine: confirm precense of<br>preCreate hook
+    IP controller-->IP controller: allocates IP address
+    IP controller->>+machine: sets IP configuration on machine
+    IP controller->>+machine: removes preCreate hook and<br>sets preTerminate hook
+    machine-->>machine controller: IP configuration read from <br>machine and converted to<br>guestinfo.afterburn.initrd.network-kargs
+    machine controller->>vCenter: creates virtual machine
+~~~
+
+
+A sample project [machine-ipam-controller](https://github.com/rvanderp3/machine-ipam-controller) is an example of a controller that implements this workflow.
+
+
+#### Variation [optional]
+
+### API Extensions
+
+The CRDs `machines.machine.openshift.io` and `machinesets.machine.openshift.io` will be modified to add a new lifecycle hook called `preCreate`.  When defined, the `preCreate` lifecycle hook will block the rendering of a machine in it's backing infrastructure.
+
+### Implementation Details/Notes/Constraints [optional]
+
+### Risks and Mitigations
+
+### Drawbacks
+
+- Scaling nodes will become more complex. This will require the OpenShift administrator to integrate IP configuration
+  management to enable scaling of machine API machine resources.
+
+- If a `machineset` is configured to specify the `preCreate` lifecycle hook, a controller must remove the hook before
+machine creation will continue.
+
+- `install-config.yaml` will grow in complexity.
+
+## Design Details
+
+### Open Questions [optional]
+
+#### `nmstate` API
+How should we introduce `nmstate` to the OpenShift API?  While we only need a subset of `nmstate` for this enhancement, `nmstate` may have broader applicability outside of vSphere.
+
+### Test Plan
+
+### Graduation Criteria
+
+#### Dev Preview -> Tech Preview
+
+- Ability to utilize the enhancement end to end
+- End user documentation, relative API stability
+- Sufficient test coverage
+- Gather feedback from users rather than just developers
+- Enumerate service level indicators (SLIs), expose SLIs as metrics
+- Write symptoms-based alerts for the component(s)
+
+#### Tech Preview -> GA
+
+- More testing (upgrade, downgrade, scale)
+- Sufficient time for feedback
+- Available by default
+- Backhaul SLI telemetry
+- Document SLOs for the component
+- Conduct load testing
+- User facing documentation created in [openshift-docs](https://github.com/openshift/openshift-docs/)
+
+**For non-optional features moving to GA, the graduation criteria must include
+end to end tests.**
+
+#### Removing a deprecated feature
+
+### Upgrade / Downgrade Strategy
+
+### Version Skew Strategy
+
+### Operational Aspects of API Extensions
+
+#### Failure Modes
+
+#### Support Procedures
+
+## Implementation History
+
+## Alternatives
```

Comment:
i'm curious if we have done any exploration of the IPAM controller for CCMs, i wonder if this might be another way for us to achieve static IPs in a more systemic way that could potentially reach other providers as well. i haven't done much exploration around the IPAM controller as we don't use, but i wonder about its inclusion on something like this enhancement.

Reply:
That is a great question @elmiko.  When https://github.com/kubernetes-sigs/cluster-api/pull/7647 was opened to explore getting a preProvision hook created upstream, the reviewer left some comments to resources that looked really interesting.  For vSphere in particular, it looks like [capv](https://github.com/kubernetes-sigs/cluster-api-provider-vsphere/blob/a93e482a6929d9f6de5fc7c7801c2b6e51e70aac/controllers/vspherevm_controller.go#L447) just waits for an IP to be added to the machine when dhcp4/6 == Disabled so presumably, some other IPAM controller could come in an assign an IP.  

And it sounds like there is some [discussion](https://github.com/kubernetes-sigs/cluster-api/blob/main/docs/proposals/20220125-ipam-integration.md) around IPAM and CAPV as well but presently that doesn't look to be implemented.

Reply:
If there's prior art upstream for a method for assigning static IPs, is it possible we can either complete the work upstream then copy downstream, or if it's already complete, copy the mechanism?